src/access/examples/ExampleAccessControlDelegate.cpp

```diff
@@ -0,0 +1,1101 @@
+                        *index = 0;
+                        for (auto & storage2 : acl)
+                        {
+                            if (&storage2 == &storage)
+                            {
+                                break;
+                            }
+                            else if (storage2.mFabricIndex == storage.mFabricIndex)
+                            {
+                                (*index)++;
+                            }
+                        }
+                    }
+                    else
+                    {
+                        *index = &storage - acl;
+                    }
+                }
+                return &storage;
+            }
+        }
+        return nullptr;
+    }
+
+    static EntryStorage * FindUsedInAcl(size_t index, const FabricIndex * fabricIndex)
+    {
+        index = FilterIndex(index, fabricIndex);
+        if (index < ArraySize(acl))
+        {
+            auto storage = acl + index;
+            if (storage->InUse())
+            {
+                return storage;
+            }
+        }
+        return nullptr;
+    }
+
+public:
+    // Pool support
+    static EntryStorage pool[kEntryStoragePoolSize];
+
+    static EntryStorage * Find(EntryStorage * candidate)
+    {
+        if (candidate && candidate->InPool())
+        {
+            return candidate;
+        }
+        for (auto & storage : pool)
+        {
+            if (!storage.InUse())
+            {
+                return &storage;
+            }
+        }
+        return nullptr;
+    }
+
+    bool InPool() const
+    {
+        constexpr auto end = pool + ArraySize(pool);
+        return pool <= this && this < end;
+    }
+
+public:
+    EntryStorage() = default;
+
+    void Init()
+    {
+        if (!mInUse)
+        {
+            Clear();
+            mInUse = true;
+        }
+    }
+
+    bool InUse() const { return mInUse; }
+
+    void Release()
+    {
+        if (InPool())
+        {
+            mInUse = false;
+        }
+    }
+
+    void Clear()
+    {
+        mInUse = false;
+        mFabricIndex = chip::kUndefinedFabricIndex;
+        mAuthMode = AuthMode::kPase;
+        mPrivilege = Privilege::kView;
+        memset(mSubjects, SubjectStorage::kEmptyByte, sizeof(mSubjects));
+#if defined(__GNUC__)
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wclass-memaccess"
+        memset(mTargets, TargetStorage::kEmptyByte, sizeof(mTargets));
+#pragma GCC diagnostic pop
+#else
+        for (auto & target : mTargets)
+        {
+            target.Clear();
+        }
+#endif
+    }
+
+private:
+    static size_t FilterIndex(size_t index, const FabricIndex * fabricIndex)
+    {
+        if (fabricIndex != nullptr)
+        {
+            size_t unfiltered = 0;
+            size_t filtered = 0;
+            for (auto & storage : acl)
+            {
+                if (storage.InUse())
+                {
+                    return ArraySize(acl);
+                }
```

Comment:
It's not obvious what this precheck does.

Reply:
It should be NOT in use. Still need to write the tests for this last bit of code. I can write comments on what's happening with these nested loops too.

Reply:
I've written the tests, and added some comments on how this function works, refactored/renamed to make it a lot clearer.

Reply:
I rewrote this function so it's clear what it's converting "from" and "to", and which is "absolute" vs. "relative" index. It should be obvious now, resolve if agree.